Automattic/wp-calypso
# Add memcached to the calypso server

### Proposed Changes
Add memcached to the Calypso server:

- Using the [`memjs`](https://www.npmjs.com/package/memjs) library. There are no canonical node libraries for memcached and dozens of random implementations. The [most popular one](https://www.npmjs.com/package/memcached) hasn't been updated in 7 years. The one I found [most recommended on the internet](https://www.npmjs.com/package/mc) is deprecated. The package with [the most recent updates](https://www.npmjs.com/package/memcache-plus) is maintained by an individual. Memjs, while not having super recent updates, is maintained by an organization that sells memcached services (I think), and the org has recent activity in general. It's also more popular than memcache-plus. The memcache-plus derivatives (like memcache-ppp) also don't inspire confidence!
- Adding a small wrapper lib in the calypso server to support typescript and to allow us to easily change the driver in the future.

### Testing Instructions

This will be tricky to test before deploying, because we ultimately need to test the production connection to the memcached server. We should probably create like one sample cache set/get and try it out on horizon.

### TODO:
- [ ] Add the memcached server address to a secret file somewhere
- [ ] Look into whether we need additional disconnect/reconnect options, and what to do if the server never exists. Should we fall back to LRU in that scenario?
- [x] Should we add some timeouts in the case that a cache get takes too long? We wouldn't want it to block a request.
- [ ] Write a small test feature that uses memcached.
